emulator/src/kreator/preprocessor.rs

```diff
@@ -180,18 +189,162 @@ fn replace_macros(code: &Vec<String>) -> Result<Vec<String>, &'static str> {
+            }
+            if !found_set_names.contains(&name.to_string()) {
+                found_set_names.push(name.to_string());
+                all_existing_names.push(name.to_string());
+            }
+        }
+
+        // change line of macro according to principles of locality
+        if in_macro {
+            
+            // map local labels in macros
+            if loc_label_regex.is_match(&owned_line) && !glob_label_regex.is_match(&owned_line) {
+                let (label, _) = owned_line.split_once(":").unwrap();
+                let gen_name = generate_label_name(&all_existing_names, &generated_label_count).unwrap();
+                generated_label_count += 1;
+                all_existing_names.push(gen_name.clone());
+                label_map.insert(label.to_string(), gen_name);
+            }
+
+            // convert globally declared label to normal label
+            if glob_label_regex.is_match(&owned_line) {
+                owned_line = owned_line.replace("::", ":");
+            }
+
+            // map local equ assignments
+            if line.contains(" EQU ") {
+                let (name, _) = owned_line.split_once(" EQU ").unwrap();
+                let gen_name = generate_label_name(&all_existing_names, &generated_label_count).unwrap();
+                generated_label_count += 1;
+                all_existing_names.push(gen_name.clone());
+                equ_map.insert(name.to_string(), gen_name);
+            }
+
+            // replace local label calls and equ assigned variables
+            for (old_label, generated_label) in &label_map {
+                owned_line = owned_line.replace(old_label, generated_label);
+            }
+            for (old_name, new_name) in &equ_map {
+                owned_line = owned_line.replace(old_name, new_name);
+            }
+            
+            // check if set assignment variable existed outside of macro
+            if owned_line.contains(" SET ") {
+                let (name, _) = owned_line.split_once(" SET ").unwrap();
+                if !found_set_names.contains(&name.to_string()) {
+                    let gen_name = generate_label_name(&all_existing_names, &generated_label_count).unwrap();
+                    generated_label_count += 1;
+                    all_existing_names.push(gen_name.clone());
+                    set_map.insert(name.to_string(), gen_name);
+                }
+            }
+            for (old_name, new_name) in &set_map {
+                owned_line = owned_line.replace(old_name, new_name);
+            }
+        }
+        handled_code.push(owned_line.to_string());
+    }
+    Ok(handled_code)
+}
+
+fn generate_label_name(taken_names: &Vec<String>, generated_label_count: &u32) -> Result<String, &'static str> {
+    let mut label_char_val = (generated_label_count / 10000) as u32 + 'A' as u32;
+    let mut label_num_val = (generated_label_count % 10000) as u32 - generated_label_count;
```

Comment:
This `- generated_label_count` should not be necessary and it looks wrong to me.

`(x % y) - x == -(x / y)`, no?

What I had in mind would mean moving the calculation of the character and the number inside the loop and incrementing `generated_label_count` each time a collision is found